src/histbuf.rs

```diff
@@ -0,0 +1,325 @@
+        p2.iter_mut().chain(p1)
+    }
+}
+
+pub struct Iter<'a, T> {
+    data: &'a [T],
+    cur: usize,
+    left: usize,
+}
+
+impl<'a, T> Iterator for Iter<'a, T> {
+    type Item = &'a T;
+
+    fn next(&mut self) -> Option<&'a T> {
+        if self.left == 0 {
+            None
+        } else {
+            let el = &self.data[self.cur];
+            self.cur = (self.cur + 1) % self.data.len();
+            self.left -= 1;
+            Some(el)
+        }
+    }
+}
+
+impl<T, N> Extend<T> for HistoryBuffer<T, N>
+where
+    N: ArrayLength<T>,
+{
+    fn extend<I>(&mut self, iter: I)
+    where
+        I: IntoIterator<Item = T>,
+    {
+        for item in iter.into_iter() {
+            self.push(item);
+        }
+    }
+}
+
+impl<'a, T, N> Extend<&'a T> for HistoryBuffer<T, N>
+where
+    T: 'a + Clone,
+    N: ArrayLength<T>,
+{
+    fn extend<I>(&mut self, iter: I)
+    where
+        I: IntoIterator<Item = &'a T>,
+    {
+        self.extend(iter.into_iter().cloned())
+    }
+}
+
+impl<T, N> AsSlice for HistoryBuffer<T, N>
+where
+    N: ArrayLength<T>,
+{
+    type Element = T;
+
+    /// Returns the array slice backing the buffer, without keeping track
+    /// of the write position. Therefore, the element order is unspecified.
+    fn as_slice(&self) -> &[T] {
+        &self.data
+    }
+}
+
+impl<T, N> AsMutSlice for HistoryBuffer<T, N>
+where
+    N: ArrayLength<T>,
+{
+    /// Returns the array slice backing the buffer, without keeping track
+    /// of the write position. Therefore, the element order is unspecified.
+    fn as_mut_slice(&mut self) -> &mut [T] {
+        &mut self.data
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::{consts::*, HistoryBuffer};
+    use as_slice::{AsSlice, AsMutSlice};
+
+    #[test]
+    fn new() {
+        let x: HistoryBuffer<u8, U4> = HistoryBuffer::new_with(1);
+        assert!(x.iter().eq([1; 4].iter()));
+
+        let x: HistoryBuffer<u8, U4> = HistoryBuffer::new();
+        assert!(x.iter().eq([0; 4].iter()));
+    }
+
+    #[test]
+    fn push_iter() {
+        let mut x: HistoryBuffer<u8, U4> = HistoryBuffer::new();
+        x.push(1);
+        x.push(4);
+        assert!(x.iter().eq([0, 0, 1, 4].iter()));
+
+        x.push(5);
+        x.push(6);
+        x.push(10);
+        assert!(x.iter().eq([4, 5, 6, 10].iter()));
+
+        x.extend([11, 12].iter());
+        assert!(x.iter().eq([6, 10, 11, 12].iter()));
+
+        assert!(x.iter_mut().eq([6, 10, 11, 12].iter()));
+    }
+
+    #[test]
+    fn clear() {
+        let mut x: HistoryBuffer<u8, U4> = HistoryBuffer::new_with(1);
+        x.clear();
+        assert!(x.iter().eq([0; 4].iter()));
+
+        let mut x: HistoryBuffer<u8, U4> = HistoryBuffer::new();
+        x.clear_with(1);
+        assert!(x.iter().eq([1; 4].iter()));
+    }
+
+    #[test]
+    fn first_last() {
+        let mut x: HistoryBuffer<u8, U4> = HistoryBuffer::new();
+        x.push(1);
+        x.push(4);
+        assert_eq!(x.first(), &0);
```

Comment:
Again, I would have expected this to return `&1` but perhaps it's OK for the use cases of this data structure.

Reply:
Thinking about this a bit more. `first` and `last` may not always be true because one can do something like `hb.as_mut_slice().sort()` or reorder the elements arbitrarily making `first` and `last` not return the original insertion order. We may want to point this out in the docs.

Reply:
Indeed; either the "mut slice" method can be removed or this has to be documented.